File: sked-android/app/src/main/kotlin/com/sked/sked_app/TimetableParser.kt
package com.sked.sked_app

import android.content.Context
import android.content.SharedPreferences
import com.sked.sked_app.widget.KEY_TODAY
import com.sked.sked_app.widget.PREFS_NAME
import org.json.JSONArray
import org.json.JSONObject
import java.util.Calendar

/**
 * On-device timetable parser — Kotlin port of sked-backend/src/timetableParser.js
 *
 * Parses the HTML fragment returned by the ASP.NET PageMethod:
 *   POST https://ums.lpu.in/lpuums/frmMyCurrentTimeTable.aspx/GetTimeTable
 *
 * Response shape: { "d": "<html string>" }
 *
 * Each class block has an onclick like:
 *   openPopup("Lecture / G:All C:INT257 / R: 33-612 / S:K2P24RL / Teacher: 30767::Akash Pundir",
 *             "10:20-11:10", "Monday", "33-612", "INT257", "L")
 *
 * Args (0-indexed):
 *   0 — full description string
 *   1 — time range  "HH:MM-HH:MM"
 *   2 — day name    e.g. "Monday"
 *   3 — room        e.g. "33-612"
 *   4 — course code e.g. "INT257"
 *   5 — class type  "L" = Lecture, "P" = Practical, "T" = Tutorial
 */
object TimetableParser {

    // ── Type mapping ─────────────────────────────────────────────────────────

    private val TYPE_LABELS = mapOf("L" to "Lecture", "P" to "Practical", "T" to "Tutorial")

    private val DAY_NAMES = arrayOf(
        "Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"
    )

    // ── Regex patterns ───────────────────────────────────────────────────────

    /**
     * Matches: openPopup("arg0","arg1","arg2","arg3","arg4","arg5")
     * Each arg is captured as a group. Handles escaped quotes inside args.
     */
    private val OPEN_POPUP_REGEX = Regex(
        """openPopup\(\s*"((?:[^"\\]|\\.)*)"\s*,\s*"((?:[^"\\]|\\.)*)"\s*,\s*"((?:[^"\\]|\\.)*)"\s*,\s*"((?:[^"\\]|\\.)*)"\s*,\s*"((?:[^"\\]|\\.)*)"\s*,\s*"((?:[^"\\]|\\.)*)"\s*\)"""
    )

    private val TEACHER_REGEX = Regex("""Teacher:\s*\d+::(.+)$""", RegexOption.IGNORE_CASE)
    private val SECTION_REGEX = Regex("""S:([^\s/]+)""")
    private val GROUP_REGEX = Regex("""G:([^\s/]+)""")

    // ── Parsing helpers ──────────────────────────────────────────────────────

    private fun parseTeacher(description: String): String {
        return TEACHER_REGEX.find(description)?.groupValues?.get(1)?.trim() ?: "Unknown"
    }

    private fun parseSection(description: String): String {
        return SECTION_REGEX.find(description)?.groupValues?.get(1)?.trim() ?: ""
    }

    private fun parseGroup(description: String): String {
        return GROUP_REGEX.find(description)?.groupValues?.get(1)?.trim() ?: ""
    }

    private fun parseTimeRange(range: String): Pair<String, String> {
        val parts = range.split("-")
        return Pair(parts.getOrElse(0) { "" }.trim(), parts.getOrElse(1) { "" }.trim())
    }

    private fun parseType(code: String): String {
        return TYPE_LABELS[code] ?: code
    }

    fun todayName(): String {
        return DAY_NAMES[Calendar.getInstance().get(Calendar.DAY_OF_WEEK) - 1]
    }

    // ── Main parsing ─────────────────────────────────────────────────────────

    /**
     * Parse the raw response from GetTimeTable (JSON string or HTML).
     * Returns a list of ClassItem entries for the entire week.
     *
     * @param rawResponse The raw response — can be:
     *   - JSON string: {"d": "<html>"}
     *   - Just the HTML string
     *   - Double-encoded JSON string from WebView
     */
    fun parse(rawResponse: String): List<ClassItem> {
        val html = extractHtml(rawResponse)
        if (html.isBlank()) return emptyList()

        // Unescape HTML entities so openPopup(&quot;...&quot;) becomes openPopup("...")
        val normalized = html
            .replace("&quot;", "\"")
            .replace("&#39;", "'")
            .replace("&amp;", "&")

        val entries = mutableListOf<ClassItem>()

        // Find all openPopup(...) calls in the HTML
        OPEN_POPUP_REGEX.findAll(normalized).forEach { match ->
            val (description, timeRange, day, room, courseCode, typeCode) = match.destructured
            val (start, end) = parseTimeRange(timeRange)

            entries.add(
                ClassItem(
                    day = day.trim(),
                    timeRange = timeRange.trim(),
                    start = start,
                    end = end,
                    room = room.trim(),
                    courseCode = courseCode.trim(),
                    type = parseType(typeCode.trim()),
                    teacher = parseTeacher(description),
                    section = parseSection(description),
                    group = parseGroup(description),
                    description = description.trim()
                )
            )
        }

        return entries
    }

    /**
     * Extract the HTML string from various possible response formats.
     */
    private fun extractHtml(rawResponse: String): String {
        var trimmed = rawResponse.trim()

        // Handle double-encoded JSON strings from WebView evaluateJavascript
        if (trimmed.startsWith("\"") && trimmed.endsWith("\"")) {
            try {
                trimmed = JSONObject("{\"v\":$trimmed}").getString("v").trim()
            } catch (_: Exception) {}
        }

        // If it starts with '<', it's already HTML
        if (trimmed.startsWith("<")) return trimmed

        // Try to parse as JSON
        return try {
            val obj = JSONObject(trimmed)
            // Standard ASP.NET PageMethod response: { "d": "<html>" }
            if (obj.has("d")) {
                obj.getString("d")
            } else {
                trimmed
            }
        } catch (_: Exception) {
            // Maybe it's a plain string — use as-is
            trimmed
        }
    }

    // ── Filtering & grouping ─────────────────────────────────────────────────

    /**
     * Filter entries for a specific day (case-insensitive), sorted by start time.
     */
    fun filterByDay(entries: List<ClassItem>, dayName: String): List<ClassItem> {
        return entries
            .filter { it.day.equals(dayName, ignoreCase = true) }
            .sortedBy { it.start }
    }

    /**
     * Group all entries by day name. Each day's list is sorted by start time.
     */
    fun groupByDay(entries: List<ClassItem>): Map<String, List<ClassItem>> {
        return entries
            .groupBy { it.day }
            .mapValues { (_, dayEntries) -> dayEntries.sortedBy { it.start } }
    }

    // ── SharedPreferences persistence ────────────────────────────────────────

    private const val KEY_ALL_ENTRIES = "all_entries"
    private const val KEY_USER_ID = "sked_user_id"

    /**
     * Save parsed timetable entries to SharedPreferences.
     * Stores both the full week and today's entries for the widget.
     */
    fun saveToPrefs(context: Context, entries: List<ClassItem>, userId: String) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

        // Save the full week as JSON array
        val allJson = entriesToJsonArray(entries)
        // Save today's entries for the widget
        val todayEntries = filterByDay(entries, todayName())
        val todayJson = entriesToJsonArray(todayEntries)

        prefs.edit()
            .putString(KEY_ALL_ENTRIES, allJson.toString())
            .putString(KEY_TODAY, todayJson.toString())
            .putString(KEY_USER_ID, userId)
            .apply()
    }

    /**
     * Load all entries from SharedPreferences.
     */
    fun loadFromPrefs(context: Context): List<ClassItem> {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val raw = prefs.getString(KEY_ALL_ENTRIES, null) ?: return emptyList()
        return jsonArrayToEntries(raw)
    }

    /**
     * Load today's entries from SharedPreferences.
     */
    fun loadTodayFromPrefs(context: Context): List<ClassItem> {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val raw = prefs.getString(KEY_TODAY, null) ?: return emptyList()
        return jsonArrayToEntries(raw)
    }

    // ── JSON serialization ───────────────────────────────────────────────────

    fun entriesToJsonArray(entries: List<ClassItem>): JSONArray {
        val arr = JSONArray()
        entries.forEach { item ->
            arr.put(JSONObject().apply {
                put("day", item.day)
                put("timeRange", item.timeRange)
                put("start", item.start)
                put("end", item.end)
                put("room", item.room)
                put("courseCode", item.courseCode)
                put("type", item.type)
                put("teacher", item.teacher)
                put("section", item.section)
                put("group", item.group)
                put("description", item.description)
            })
        }
        return arr
    }

    private fun jsonArrayToEntries(raw: String): List<ClassItem> {
        return try {
            val arr = JSONArray(raw)
            (0 until arr.length()).map { i ->
                val obj = arr.getJSONObject(i)
                ClassItem(
                    day = obj.optString("day"),
                    timeRange = obj.optString("timeRange"),
                    start = obj.optString("start"),
                    end = obj.optString("end"),
                    room = obj.optString("room"),
                    courseCode = obj.optString("courseCode"),
                    type = obj.optString("type", "Lecture"),
                    teacher = obj.optString("teacher"),
                    section = obj.optString("section"),
                    group = obj.optString("group"),
                    description = obj.optString("description")
                )
            }
        } catch (_: Exception) {
            emptyList()
        }
    }
}

File: sked-app/lib/services/timetable_parser.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/class_entry.dart';

/// On-device timetable parser — Dart port of TimetableParser.kt.
///
/// Parses the HTML fragment returned by the ASP.NET PageMethod:
///   POST https://ums.lpu.in/lpuums/frmMyCurrentTimeTable.aspx/GetTimeTable
class TimetableParser {
  static const String prefsAllEntries = 'all_entries';
  static const String prefsToday = 'sked_today_entries';
  static const String prefsUserId = 'sked_user_id';

  static const Map<String, String> typeLabels = {
    'L': 'Lecture',
    'P': 'Practical',
    'T': 'Tutorial',
  };

  static const List<String> dayNames = [
    'Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'
  ];

  static final RegExp _openPopupRegex = RegExp(
    r'openPopup\(\s*"((?:[^"\\]|\\.)*)"\s*,\s*"((?:[^"\\]|\\.)*)"\s*,\s*"((?:[^"\\]|\\.)*)"\s*,\s*"((?:[^"\\]|\\.)*)"\s*,\s*"((?:[^"\\]|\\.)*)"\s*,\s*"((?:[^"\\]|\\.)*)"\s*\)',
  );

  static final RegExp _teacherRegex = RegExp(r'Teacher:\s*\d+::(.+)$', caseSensitive: false);
  static final RegExp _sectionRegex = RegExp(r'S:([^\s/]+)');
  static final RegExp _groupRegex = RegExp(r'G:([^\s/]+)');

  static String parseTeacher(String description) {
    final m = _teacherRegex.firstMatch(description);
    return m?.group(1)?.trim() ?? 'Unknown';
  }

  static String parseSection(String description) {
    final m = _sectionRegex.firstMatch(description);
    return m?.group(1)?.trim() ?? '';
  }

  static String parseGroup(String description) {
    final m = _groupRegex.firstMatch(description);
    return m?.group(1)?.trim() ?? '';
  }

  static (String start, String end) parseTimeRange(String range) {
    final parts = range.split('-');
    final start = parts.isNotEmpty ? parts[0].trim() : '';
    final end = parts.length > 1 ? parts[1].trim() : '';
    return (start, end);
  }

  static String parseType(String code) {
    return typeLabels[code] ?? code;
  }

  static String todayName() {
    final dt = DateTime.now();
    // DateTime weekday: 1 = Monday, ..., 7 = Sunday
    // Map to dayNames array where 0 = Sunday, 1 = Monday, ...
    final sundayZeroIndex = dt.weekday % 7;
    return dayNames[sundayZeroIndex];
  }

  /// Parses the raw response from GetTimeTable (JSON string or HTML).
  static List<ClassEntry> parse(String rawResponse) {
    final html = extractHtml(rawResponse);
    if (html.trim().isEmpty) return [];

    // Unescape HTML entities so openPopup(&quot;...&quot;) becomes openPopup("...")
    final normalized = html
        .replaceAll('&quot;', '"')
        .replaceAll('&#39;', "'")
        .replaceAll('&amp;', '&');

    final entries = <ClassEntry>[];

    for (final match in _openPopupRegex.allMatches(normalized)) {
      final description = match.group(1) ?? '';
      final timeRange = match.group(2) ?? '';
      final day = match.group(3) ?? '';
      final room = match.group(4) ?? '';
      final courseCode = match.group(5) ?? '';
      final typeCode = match.group(6) ?? '';

      final (start, end) = parseTimeRange(timeRange);

      entries.add(
        ClassEntry(
          day: day.trim(),
          timeRange: timeRange.trim(),
          start: start,
          end: end,
          room: room.trim(),
          courseCode: courseCode.trim(),
          type: parseType(typeCode.trim()),
          teacher: parseTeacher(description),
          section: parseSection(description),
          group: parseGroup(description),
          description: description.trim(),
        ),
      );
    }

    return entries;
  }

  /// Extracts HTML string from possible response formats (JSON object, double encoded, raw HTML).
  static String extractHtml(String rawResponse) {
    var trimmed = rawResponse.trim();

    // Handle double-encoded JSON string from WebView evaluateJavascript
    if (trimmed.startsWith('"') && trimmed.endsWith('"')) {
      try {
        final decoded = jsonDecode('{"v":$trimmed}') as Map<String, dynamic>;
        trimmed = (decoded['v'] as String? ?? '').trim();
      } catch (_) {}
    }

    // If it starts with '<', it's already HTML
    if (trimmed.startsWith('<')) return trimmed;

    // Try to parse as JSON
    try {
      final decoded = jsonDecode(trimmed);
      if (decoded is Map<String, dynamic>) {
        if (decoded.containsKey('d') && decoded['d'] is String) {
          return decoded['d'] as String;
        }
      }
      return trimmed;
    } catch (_) {
      return trimmed;
    }
  }

  /// Filter entries for a specific day (case-insensitive), sorted by start time.
  static List<ClassEntry> filterByDay(List<ClassEntry> entries, String dayName) {
    final filtered = entries.where((e) => e.day.toLowerCase() == dayName.toLowerCase()).toList();
    filtered.sort((a, b) => a.start.compareTo(b.start));
    return filtered;
  }

  /// Group all entries by day name. Each day's list is sorted by start time.
  static Map<String, List<ClassEntry>> groupByDay(List<ClassEntry> entries) {
    final map = <String, List<ClassEntry>>{};
    for (final e in entries) {
      map.putIfAbsent(e.day, () => []).add(e);
    }
    for (final day in map.keys) {
      map[day]!.sort((a, b) => a.start.compareTo(b.start));
    }
    return map;
  }

  // ── SharedPreferences persistence ────────────────────────────────────────

  static Future<void> saveToPrefs(List<ClassEntry> entries, String userId) async {
    final prefs = await SharedPreferences.getInstance();
    final allJson = jsonEncode(entries.map((e) => e.toJson()).toList());
    final todayEntries = filterByDay(entries, todayName());
    final todayJson = jsonEncode(todayEntries.map((e) => e.toJson()).toList());

    await prefs.setString(prefsAllEntries, allJson);
    await prefs.setString(prefsToday, todayJson);
    await prefs.setString(prefsUserId, userId);
  }

  static Future<List<ClassEntry>> loadFromPrefs() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(prefsAllEntries);
    if (raw == null || raw.trim().isEmpty) return [];

    try {
      final decoded = jsonDecode(raw) as List<dynamic>;
      return decoded.map((e) => ClassEntry.fromJson(e as Map<String, dynamic>)).toList();
    } catch (_) {
      return [];
    }
  }

  static Future<List<ClassEntry>> loadTodayFromPrefs() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(prefsToday);
    if (raw == null || raw.trim().isEmpty) return [];

    try {
      final decoded = jsonDecode(raw) as List<dynamic>;
      return decoded.map((e) => ClassEntry.fromJson(e as Map<String, dynamic>)).toList();
    } catch (_) {
      return [];
    }
  }
}
